### ISSUE TITLE
Inconsistent Focus Tracking + z-index issues

### ISSUE THREAD

**Issue by apfelbox** [ckeditor/ckeditor5#6656]
## 📝 Provide detailed reproduction steps (if any)

I have a project, that uses a ckeditor in a longer form. I now have z-index issues, if the toolbar dropdowns are longer than the actual ckeditor itself.

I can easily solve this, if I use `:focus-within` in CSS and increase the zindex of this form row, if the focus is inside it. That works somehow, but breaks if I only click the toolbar, as the focus doesn't jump into the field yet.
Is there some existing workaround or a specific event I can listen to that adds a class to the form row when interacting with the editor? I haven't found anything that triggers when just opening a toolbar dropdown without clicking into the editor.

PS: I am using a `DecoupledEditor`.

### ✔️ Expected result

It should look something like this:

![2020-04-23 at 15 06](https://user-images.githubusercontent.com/1032411/80103138-08e2c000-8574-11ea-821a-4fd8a6f47075.png)

### ❌ Actual result

It looks something like this:

![Kapture 2020-04-23 at 15 12 21](https://user-images.githubusercontent.com/1032411/80107766-ccb35d80-857b-11ea-8c48-0933d404586d.gif)

## 📃 Other details

* Browser: Chrome 81.0.4044.122
* OS: macOS 10.15.4 (19E287)
* CKEditor version: see below
* Installed CKEditor plugins: see below

```
▲ npm ls --depth 0
@my/app@2.0.5 /Users/.../my-app
├── @ckeditor/ckeditor5-alignment@18.0.0
├── @ckeditor/ckeditor5-autoformat@18.0.0
├── @ckeditor/ckeditor5-basic-styles@18.0.0
├── @ckeditor/ckeditor5-dev-utils@13.0.0
├── @ckeditor/ckeditor5-editor-decoupled@18.0.0
├── @ckeditor/ckeditor5-essentials@18.0.0
├── @ckeditor/ckeditor5-heading@18.0.0
├── @ckeditor/ckeditor5-link@18.0.0
├── @ckeditor/ckeditor5-list@18.0.0
├── @ckeditor/ckeditor5-paragraph@18.0.0
├── @ckeditor/ckeditor5-paste-from-office@18.0.0
├── @ckeditor/ckeditor5-special-characters@18.0.0
├── @ckeditor/ckeditor5-theme-lark@18.0.0
...
```

---

If you'd like to see this fixed sooner, add a 👍 reaction to this post.

**Comment by Reinmar** [comment]
cc @oleq

**Comment by oleq** [comment]
Hi @apfelbox!

The issue you're experiencing is caused by `tabindex="-1"` set on all toolbar items in CKEditor 5 (and `preventDefault()` called on `mousedown` of each button). We are using these techniques to make sure the toolbar does not "steal" the focus from the editing root (where you type text) when being clicked.

**Comment by oleq** [comment]
Hi, @apfelbox!

The issue you're experiencing is caused by `tabindex="-1"` set on all toolbar items in CKEditor 5 (and `preventDefault()` called on `mousedown` of each button). Long story short, we are using these techniques to make sure the toolbar does not "steal" the focus from the editing root (where you type text) when being clicked.

This is why `:focus-within` does not work when you click a button if the editing root didn't have focus in the first place – the button does not receive focus and, as I mentioned, the editing root does not have focus either.

BTW: the `:focus-within` will also fail if you open a link pop-up and the focus will go to the URL input. All floating elements of CKEditor 5 are injected in DOM in a separate `.ck-body-wrapper` container at the end of `<body>` and they do not share a common ancestor with the editing root you could apply `:focus-within` to like you did.

Until we address this problem in some other way, I'd suggest you permanently increase the `z-index` of `.ck.ck-dropdown__panel`. Let us know if this solves your problem.

**Comment by oleq** [comment]
BTW, this comes down to https://github.com/ckeditor/ckeditor5/issues/2000.

**Comment by apfelbox** [comment]
Hi @oleq 

Thanks for your suggestion! I will definitely try that, although iirc my `.form-row` is a new stacking context (`position: relative`), so it probably doesn't solve the issue as my `.form-row` would need the higher z-index instead of a child inside of it.
I could increase it for every `.form-row` that contains a RTE field, but then I would have the same issue if I have two RTE fields right next to each other.

> BTW: the :focus-within will also fail if you open a link pop-up and the focus will go to the URL input. All floating elements of CKEditor 5 are injected in DOM in a separate .ck-body-wrapper container at the end of <body> and they do not share a common ancestor with the editing root you could apply :focus-within to like you did.

Ah, thanks. Yeah, that won't help there then, but at least these pop-ups normally don't extend over the RTE like a long format dropdown does.

> [...]

So there (currently?) is nothing I can listen to to temporarily set an additional class on the `.form-row`?

Something like (just pseudo code, please ignore the specifics):

```js
const toolbar = editor.ui.view.toolbar;

toolbar.on("open-dropdown", formRow.classList.add("is-focused"));
toolbar.on("close-dropdown", formRow.classList.remove("is-focused"));
```

So basically anything I can hook into to recognize a dropdown being opened / closed?

**Comment by oleq** [comment]
> So basically anything I can hook into to recognize a dropdown being opened / closed?

Until https://github.com/ckeditor/ckeditor5/issues/2000 is resolved, the only thing you could ATM that popped into my head is:

```js
const dropdownsInTheToolbar = editor.ui.view.toolbar.items
    .filter( item => {
        // Duck typing to find instances of DropdownView only.
        return item.panelView && item.buttonView;
    } );

editor.ui.view.set( 'isAnyDropdownOpen', false );
editor.ui.view.bind( 'isAnyDropdownOpen' ).toMany( dropdownsInTheToolbar, 'isOpen', ( ...areOpen ) => {
    return areOpen.some( isDropdownOpen => isDropdownOpen );
} );

editor.ui.view.on( 'change:isAnyDropdownOpen', () => {
    console.log( `isAnyDropdownOpen: ${ editor.ui.view.isAnyDropdownOpen }` );
} );
```

This isn't pretty but it should work. It finds all dropdowns in the toolbar and creates an observable `editor.ui.view.isAnyDropdownOpen` property you can listen to that will tell if any dropdown is open.

**Comment by apfelbox** [comment]
Thanks @oleq,

that works perfectly 👍 

I implemented it like this for now, but I am happy to refactor this part with a global "notify me if anything in the editor got / lost focus" observer. 😄
